### PR TITLE
[Stepper] Add Alternative Label on horizontal-type Stepper

### PR DESCRIPTION
**Pull request by KKimj**
# Add Alternative Label on horizontal-type Stepper

## Linked issues

#91337

### Before
image

---

### Horizontal

image

<del>
### Vertical

image

### [important!] Remained issues

- Padding, Align
image

- Up and down number-icon
image
</del>

## Reproduce code
```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      title: 'Stepper Alternative Label',
      home: MyStatefulWidget(),
    );
  }
}

class MyStatefulWidget extends StatefulWidget {
  const MyStatefulWidget({Key? key}) : super(key: key);

  @override
  State<MyStatefulWidget> createState() => _MyStatefulWidgetState();
}

class _MyStatefulWidgetState extends State<MyStatefulWidget> {
  int _index = 0;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Stepper(
        type: StepperType.horizontal,
        // type: StepperType.vertical,
        currentStep: _index,
        onStepCancel: () {
          if (_index > 0) {
            setState(() {
              _index -= 1;
            });
          }
        },
        onStepContinue: () {
          if (_index < 3) {
            setState(() {
              _index += 1;
            });
          }
        },
        onStepTapped: (int index) {
          setState(() {
            _index = index;
          });
        },
        steps: <Step>[
          Step(
            isActive: _index == 0,
            title: const Text('Step 1 title'),
            content: Container(
                alignment: Alignment.centerLeft,
                child: const Text('Content for Step 1')),
            label: const Text('Step 1'),
            subtitle: const Text('Step 1 subtitle'),
          ),
          Step(
            isActive: _index == 1,
            title: const Text('Step 2 title'),
            content: const Text('Content for Step 2'),
            subtitle: const Text('Step 2 subtitle'),
            label: const Text('Step 2'),
          ),
          Step(
            isActive: _index == 2,
            title: const Text('Step 3 title'),
            content: const Text('Content for Step 3'),
            subtitle: const Text('Step 3 subtitle'),
            label: const Text('Step 3'),
          ),
          Step(
            isActive: _index == 3,
            title: const Text('Step 4 title'),
            content: const Text('Content for Step 4'),
            subtitle: const Text('Step 4 subtitle'),
            label: const Text('Step 4'),
          ),
        ],
      ),
    );
  }
}
```

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by craiglabenz** [comment]
Thanks for the confirmation on tests, @KKimj. If you're new to widget tests, the existing Stepper widget tests contain many great examples of meaningful tests :)

**Comment by KKimj** [comment]
@craiglabenz 
Hello, I add new test for lable ( Alternative Label ) with stepper.
Please check, and give me some comments.
Thanks, HAGW!

**Review by craiglabenz — CHANGES_REQUESTED** [review]
Thanks so much for this PR, @KKimj :)

I left two comments, after which I think this will be good to go!

**Comment by KKimj** [comment]
@craiglabenz 
Hello, how are you doing?
Can you give me a new feedback?
Thank you.

**Comment by craiglabenz** [comment]
@KKimj 

I am *So. So.* sorry for disappearing for so long. Somehow, the alert for this PR snuck past me and I never caught it.

In my last comment, I wrote "it would be nice to add another one that confirms the styling logic you've written is behaving as anticipated", which I now realize was a little unclear. The key function I would like to see tested is `_labelStyle`, since it accepts an index and merges that with the user's progress through the stepper and the Dark Mode / Light Mode status before ultimately yielding a `TextStyle`. The internals of that function are pretty complicated, so I think we should have unit tests to verify that a Stepper with X steps, and currently at step Y, is rendering each label correctly. There a few scenarios to cover there - steps before, on, and after the active step; in both light mode and dark mode.

Does all of that make sense?

PS - I promise not to disappear this time!

**Comment by KKimj** [comment]
Hello!! LTNS!!
I understand what you say.

I will update test-codes soon.
Something like below.

- Let, `X := currentStepIndex`, `Y := arbitraryStepIndex (except current)`

1. set X = 0.
2. Check Styles of X and Y.
3. set X = 1.
4. Invoke `setState()`.
5. Check Styles of X and Y.

P.S 
`Happy New Year!`

**Comment by craiglabenz** [comment]
Yes, that looks like the right pseudocode. Thanks so much and 🌔 Happy New Year 🌔 to you, too!

**Comment by Hixie** [comment]
@KKimj Is this a PR you are interested in doing further work on? It's perfectly understandable if you don't have time right now, and if that's the case we can save what you've done so far and put it aside so that if someone else wants to work on this they can start from your PR. But if you do have the time to work on this, that would be great!

**Comment by KKimj** [comment]
@Hixie 
Hello!
I always had it in mind. I want to finish all the work within April. How does that sound?

File: packages/flutter/test/material/stepper_test.dart [hunk 1]
```diff
@@ -1156,6 +1156,56 @@ testWidgets('Stepper custom indexed controls test', (WidgetTester tester) async
...
+        ),
+      ),
+    );
+    // Tap to Step3 Label then, index become 2 from 0
+    await tester.tap(find.text('Step 3 Label'));
+    expect(index, 2);
+    // Tap to Step2 Label then, index become 1 from 2
+    await tester.tap(find.text('Step 2 Label'));
+    expect(index, 1);
+  });
```
[the remark this answers is not in the document]
Comment (KKimj) [reply packages/flutter/test/material/stepper_test.dart:1249]: @craiglabenz 
Hello..! It's been so long. How have you been?

I pushed a commit for test about `TextStyles` of Label Widget.

Please Review my commit.
Thanks!

**Comment by KKimj** [comment]
@craiglabenz 
Hello..! It's been so long. How have you been?

I pushed a commit for test about `TextStyles` of Label Widget.

Please Review my commit.
Thanks!

**Comment by craiglabenz** [comment]
That test looks good, @KKimj. However, I just tested the code myself directly, and the padding / alignment issue does still seem to be in play. Is this something you're also looking to address?

**Comment by KKimj** [comment]
@craiglabenz 
Hello!

Now I realize again that there is no alternative label in the Vertical type, `StepperType.vertical`.
- https://material.io/archive/guidelines/components/steppers.html#steppers-types-of-steppers
image

Anyway, I found a way to solve the `padding / alignment issue`.
The method is to modify the function '_buildVerticalHeader', forcing width to 24.0, according to the Material Guide.
- Code diff here, (It was not pushed.) 
- image
- An example UI can be similar to the following.
image
But it's too small, and as I said above, it's dubious because `alternative label` is not in the Vertical type. I'd like to ask for advice on how to do it.

Thanks, Have a good day!

**Comment by craiglabenz** [comment]
Oh dear, we're running out of pixels below the circle. To be honest, I'm not sure that it makes sense to put anything in this spot, given the screenshots you shared and how cramped the area is. The fact that only such short strings will fit is a bit of a problem, in my opinion.

Taking a step back - what informed your desire to land this change? Is there possibly another way that need could be satsified?

**Comment by KKimj** [comment]
@craiglabenz 
I was motivated for the horizontal type.
Is it right to exclude it from the vertical type?

Or is it right to include even more width?
But this would violate the width or margin value of the Material design.

**Comment by craiglabenz** [comment]
I think it's probably fair to have this be horizontal-only, and to spell that out in the relevant docstrings.

**Comment by KKimj** [comment]
Thank you!

Then, when the label attribute is given in the vertical type, how should I handle it?
Should I use an asset or something?

**Comment by craiglabenz** [comment]
My ideal solution would simply be to do nothing in the vertical case. Have the new attribute's docstring mention that it is only honored in the horizontal alignment and then make no changes to the vertical build methods.

**Comment by KKimj** [comment]
Thank you for letting me know about your Insight!
I just did some new push.
Please check. Thank you!

**Comment by craiglabenz** [comment]
I like that this is currently horizontal-only, but I'm also seeing a spacing issue there. Note that before adding a `label`, everything is appropriately centered:

Screen Shot 2022-04-25 at 11 13 19 AM

But then after adding a label, the horizontal connectors shift and are no-longer centered. fwiw, I think this is largely the core challenge of this PR - you can no longer as easily predict the heights of each component to know how to keep the horizontal connectors centered within the circle.

Screen Shot 2022-04-25 at 11 13 33 AM

The horizontal connectors look centered within the entire column of circle-and-label, but my hunch is that it should remain centered within the _circle_.

**Comment by KKimj** [comment]
@craiglabenz 

This is what you're expecting, right?
스크린샷 2022-04-26 오전 3 59 09

**Comment by craiglabenz** [comment]
That centering looks good! I forget, is the elevation coming from the Stepper itself, or have you wrapped it in something in this demo? If it comes from the Stepper itself, then some padding below the `label` would help.

**Comment by KKimj** [comment]
@craiglabenz 
Hello!! 
I Pushed some commits.
Please review!

image

image

image

**Comment by KKimj** [comment]
And also, Please check out these corner cases, which are related to `fontSize`.

image

**Review by craiglabenz — APPROVED** [review]
Nice, I think this is ready!

**Comment by KKimj** [comment]
@craiglabenz 
Wow, thank you.
Is there anything I need to do for `customer_testing-linux`, CI that has failed?

**Comment by craiglabenz** [comment]
I strongly suspect that failure is a false-negative, but we'll have to see if it happens again.

File: packages/flutter/lib/src/material/stepper.dart [hunk 2]
```diff
@@ -162,6 +163,9 @@ class Step {
 
   /// Whether or not the step is active. The flag only influences styling.
   final bool isActive;
+
+  /// [label] appears under the number icon.
```
[the remark this answers is not in the document]
Comment (KKimj) [reply packages/flutter/lib/src/material/stepper.dart]: @craiglabenz 
Would you change to status resolved?

File: packages/flutter/test/material/stepper_test.dart (hunk 1 above)
[the remark this answers is not in the document]
Comment (KKimj) [reply packages/flutter/test/material/stepper_test.dart:1230]: @craiglabenz 
Would you change to status resolved?
Thanks!

**Pull request merged**